File: V8-1/JT-Loader_Vis-Studio_C#/JTLoader/Codecs/DeeringNormalLookupTable.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace C_sharp_JT_Reader.Codecs
{
    public class DeeringNormalLookupTable
    {
        int nBits;
        double[] cosTheta;
        double[] sinTheta;
        double[] cosPsi;
        double[] sinPsi;

        public DeeringNormalLookupTable()
        {
            int numberbits = 8;
            nBits = Math.Min(numberbits, 31);
            int tableSize = (1 << nBits);
            cosTheta = new double[tableSize + 1];
            sinTheta = new double[tableSize + 1];
            cosPsi = new double[tableSize + 1];
            sinPsi = new double[tableSize + 1];

            double psiMax = 0.615479709;

            double fTableSize = tableSize;

            for (int ii = 0; ii <= tableSize; ii++)
            {
                double theta = Math.Asin(Math.Tan(psiMax * (tableSize - ii) / fTableSize));
                double psi = psiMax * ((ii) / fTableSize);
                cosTheta[ii] = Math.Cos(theta);
                sinTheta[ii] = Math.Sin(theta);
                cosPsi[ii] = Math.Cos(psi);
                sinPsi[ii] = Math.Sin(psi);
            }
        }

        public int numBitsPerAngle()
        {
            return nBits;
        }

        public DeeringLookupEntry lookupThetaPsi(long theta, long psi, int numberBits)
        {
            int offset = nBits - numberBits;

            long offTheta = (theta << offset) & 0xFFFFFFFFL;
            long offPsi = (psi << offset) & 0xFFFFFFFFL;

            return new DeeringLookupEntry(cosTheta[(int)offTheta],
                    sinTheta[(int)offTheta], cosPsi[(int)offPsi],
                    sinPsi[(int)offPsi]);
        }
    }
}


/*




*/

File: V8-1/JT-Loader_VSCode_TS/src/DeeringNormalLookupTable.ts
import { DeeringLookupEntry } from "./DeeringLookupEntry.js";

export class DeeringNormalLookupTable {
    nBits: number;
    cosTheta: number[];
    sinTheta: number[];
    cosPsi: number[];
    sinPsi: number[];

    constructor() {
        var numberbits: number = 8;
        this.nBits = Math.min(numberbits, 31);
        var tableSize: number = (1 << this.nBits);
        this.cosTheta = new Array(tableSize + 1);
        this.sinTheta = new Array(tableSize + 1);
        this.cosPsi = new Array(tableSize + 1);
        this.sinPsi = new Array(tableSize + 1);
        var psiMax: number = 0.615479709;
        var fTableSize: number = tableSize;
        for (var ii: number = 0; ii <= tableSize; ii++) {
            var theta: number = Math.asin(Math.tan(psiMax * (tableSize - ii) / fTableSize));
            var psi: number = psiMax * ((ii) / fTableSize);
            this.cosTheta[ii] = Math.cos(theta);
            this.sinTheta[ii] = Math.sin(theta);
            this.cosPsi[ii] = Math.cos(psi);
            this.sinPsi[ii] = Math.sin(psi);
        }
    }

    public numBitsPerAngle(): number {
        return this.nBits;
    }

    public lookupThetaPsi(theta: number, psi: number, numberBits: number): DeeringLookupEntry {
        var offset: number = this.nBits - numberBits;
        var offTheta: number = (theta << offset) & 0xFFFFFFFF;
        var offPsi: number = (psi << offset) & 0xFFFFFFFF;
        return new DeeringLookupEntry(this.cosTheta[<number>offTheta],
            this.sinTheta[<number>offTheta], this.cosPsi[<number>offPsi],
            this.sinPsi[<number>offPsi]);
    }
}
